crates/duty-tracker/src/contract_manager.rs

```diff
@@ -972,51 +893,45 @@ impl ContractManagerCtx {
+            } => {
                 let txid = Txid::from_byte_array(*session_id.as_ref());
 
-                if let Some(deposit_txid) = self.state.claim_txids.get(&txid) {
-                    if let Some(contract) = self.state.active_contracts.get_actor(deposit_txid) {
-                        let claim_txid = txid;
-                        duties.extend(
-                            contract
-                                .process_event(ContractEvent::GraphNonces {
-                                    signer: key,
-                                    claim_txid,
-                                    pubnonces: nonces,
-                                })
-                                .await?,
-                        );
-                    }
-                } else {
-                    // Use synchronous index-based lookup to avoid deadlock.
-                    if let Some(contract) = self
-                        .state
-                        .active_contracts
-                        .get_actor_by_deposit_request_txid(&txid)
-                    {
-                        if nonces.len() != 1 {
-                            return Err(ContractManagerErr::InvalidP2PMessage(Box::new(
-                                UnsignedGossipsubMsg::Musig2NoncesExchange { session_id, nonces },
-                            )));
-                        }
-                        let nonce = nonces.into_iter().next().unwrap();
-                        duties.extend(
-                            contract
-                                .process_event(ContractEvent::RootNonce(key.clone(), nonce))
-                                .await?,
-                        );
-                    } else {
-                        warn!(txid=%txid, "received nonces exchange for unknown session - contract may not be indexed yet");
+                if let Some(contract) = self
+                    .state
+                    .claim_txids
+                    .get(&txid)
+                    .and_then(|deposit_txid| self.state.active_contracts.get_mut(deposit_txid))
+                {
+                    let claim_txid = txid;
+                    duties.extend(contract.process_contract_event(ContractEvent::GraphNonces {
+                        signer: key,
+                        claim_txid,
+                        pubnonces: nonces,
+                    })?);
+                } else if let Some((_, contract)) = self
+                    .state
+                    .active_contracts
+                    .iter_mut()
+                    .find(|(_, contract)| contract.deposit_request_txid() == txid)
+                {
+                    if nonces.len() != 1 {
+                        return Err(ContractManagerErr::InvalidP2PMessage(Box::new(
+                            UnsignedGossipsubMsg::Musig2NoncesExchange { session_id, nonces },
+                        )));
                     }
+                    let nonce = nonces.pop().unwrap();
```

Comment:
combine these by `let Some(nonce) = nonces.pop() else {...}`

Reply:
Done in 030f0acaf57b917f97921033caf6161e2f52f0a5